Sources/CacheAdvance/CacheReader.swift

```diff
@@ -60,15 +75,16 @@ final class CacheReader {
             return message
 
         case .emptyRead:
-            guard !previousReadWasEmpty else {
-                // If the previous read was also empty, then the file has been corrupted.
+            guard !(startingOffset < offsetInFileAtEndOfNewestMessage) else {
```

Comment:
```suggestion
            guard startingOffset >= offsetInFileAtEndOfNewestMessage else {
```

Reply:
In this condition, so we will seek to `FileHeader.expectedEndOfHeaderInFile` when `startingOffset >= offsetInFileAtEndOfNewestMessage`. Why do we need to go back? Is it related with `overwritesOldMessages == true`?

Reply:
I went back and forth on whether to use `!` here – I appreciate your weighing in. I'll change it 🙂

We check `startingOffset >= offsetInFileAtEndOfNewestMessage` here because if `startingOffset < offsetInFileAtEndOfNewestMessage`, we expect to be reading a message! If we read empty data under this condition, we know that our `offsetInFileAtEndOfNewestMessage` is incorrect, or that a message we read earlier wrote an incorrect message length – i.e. we know that the file is corrupted.

This check is implicitly related to `overwritesOldMessages == true`. As we discussed in #64, when `overwritesOldMessages == false`, we should never seek back to the beginning of the file. On this PR, when `overwritesOldMessages == false` we will not seek back due to a corrupted file: either this check will catch a corrupted file, or this check will:
https://github.com/dfed/CacheAdvance/blob/4d6b753ae2b9bbd837dd7a03ddab50b4b4b0eac0/Sources/CacheAdvance/CacheReader.swift

You can prove the above to yourself by setting `overwritesOldMessages: false` in the unit tests I wrote on this PR.

Reply:
Got it. Now I got fully understanding for `overwritesOldMessages == true` : )

Reply:
Oops! Updated above, and linking here:
https://github.com/dfed/CacheAdvance/blob/fd03b6d0d6584e8a66c90ca77acbbf4db23d50bc/Sources/CacheAdvance/CacheReader.swift#L58-L68